src/Sentry.AspNetCore/SentryTracingBuilder.cs

```diff
@@ -0,0 +1,61 @@
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Sentry.AspNetCore;
+
+/// <summary>
+/// Delegates/wraps the inner builder, so that we can intercept calls to <see cref="Use"/> and add our middleware
+/// </summary>
+internal class SentryTracingBuilder : IApplicationBuilder
+{
+    private const string EndpointRouteBuilder = "__EndpointRouteBuilder";
+    public SentryTracingBuilder(IApplicationBuilder inner)
+    {
+        InnerBuilder = inner;
+    }
+
+    private IApplicationBuilder InnerBuilder { get; }
+
+    /// <inheritdoc />
+    public IServiceProvider ApplicationServices
+    {
+        get => InnerBuilder.ApplicationServices;
+        set => InnerBuilder.ApplicationServices = value;
+    }
+
+    /// <inheritdoc />
+    public IDictionary<string, object?> Properties => InnerBuilder.Properties;
+
+    /// <inheritdoc />
+    public IFeatureCollection ServerFeatures => InnerBuilder.ServerFeatures;
+
+    /// <inheritdoc />
+    public RequestDelegate Build() => InnerBuilder.Build();
+
+    /// <inheritdoc />
+    public IApplicationBuilder New() => new SentryTracingBuilder(InnerBuilder.New());
+
+    /// <inheritdoc />
+    public IApplicationBuilder Use(Func<RequestDelegate, RequestDelegate> middleware)
+    {
+        // UseRouting sets a property on the builder that we use to detect when UseRouting has been added and we
+        // register SentryTracing immediately afterwards
+        // https://github.com/dotnet/aspnetcore/blob/8eaf4b51f73ae2b0ed079e4f8253afb53e96b703/src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs#L58-L62
+        if (Properties.ContainsKey(EndpointRouteBuilder) && this.ShouldRegisterSentryTracing())
```

Comment:
Do I get that correctly: Since the `UseRouting` sets a flag and we check for that flag before forwarding to the InnerBuilder, Sentry will add the `TracingMiddleware` with the `UseX` that follows after `UseRouting`?

Reply:
Yeah this is what happens in the dotnet repo:

https://github.com/dotnet/aspnetcore/blob/8eaf4b51f73ae2b0ed079e4f8253afb53e96b703/src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs#L58-L62

You can see they set that property immediately before calling `UseMiddleware` (which is then intercepted by `SentryTracingBuilder`). As soon as we see this property on the builder then, we know a call has been made to `UseRouting` and we can safely register our tracing middleware. 

It shouldn't matter whether something comes after UseRouting or not... the way it sits at the moment, we would detect the call to UseRouting, the property would already have been set at that stage and we'd register both the UseRouting middleware and the UseSentryTracing middleware. If anything gets registered after UseRouting or not doesn't matter much then.

Initially I was making the call to `InnerBuilder.Use(middleware)` before checking for the existence of the property... just in case Microsoft ever changed the order in which they made the call to UseMiddleware and set the build property. However that would result in an extra (potentially unecessary) operation.